dspy/functional/typed_predictor_signature.py

```diff
@@ -0,0 +1,130 @@
+from typing import Annotated, Type, Union
+from typing import get_origin, get_args
+
+from pydantic import BaseModel
+from pydantic_core import PydanticUndefined
+from pydantic.fields import FieldInfo
+
+import dspy
+from dspy import InputField, OutputField, Signature
+
+
+class TypedPredictorSignature:
+    @classmethod
+    def create(
+            cls,
+            pydantic_class_for_dspy_input_fields: Type[BaseModel],
+            pydantic_class_for_dspy_output_fields: Type[BaseModel],
+            prefix_instructions: str = "") -> Type[Signature]:
+        """
+        Return a DSPy Signature class that can be used to extract the output parameters.
+
+        :param pydantic_class_for_dspy_input_fields: Pydantic class that defines the DSPy InputField's.
+        :param pydantic_class_for_dspy_output_fields: Pydantic class that defines the DSPy OutputField's.
+        :param prefix_instructions: Optional text that is prefixed to the instructions.
+        :return: A DSPy Signature class optimizedfor use with a TypedPredictor to extract structured information.
+        """
+        if prefix_instructions:
+            prefix_instructions += "\n\n"
+        instructions = prefix_instructions + "Use only the available information to extract the output fields.\n\n"
+        dspy_fields = {}
+        for field_name, field in pydantic_class_for_dspy_input_fields.model_fields.items():
+            if field.default and 'typing.Annotated' in str(field.default):
+                raise ValueError(f"Field '{field_name}' is annotated incorrectly. See 'Constraints on compound types' in https://docs.pydantic.dev/latest/concepts/fields/")
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
```

Comment:
interesting to see examples and metadata per field

Reply:
these are definitely valuable signals. We should leverage Pydantic field metadata instead of duplicating in DSPy signature fields

Reply:
Update: I am no longer using examples. It turns out LLM's will incorrectly return an example as the parameter value instead of "NOT_FOUND"